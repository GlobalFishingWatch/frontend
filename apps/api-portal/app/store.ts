import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit'
import userApplicationsReducer from 'features/user-applications/user-applications.slice'
import userReducer from '../features/user/user.slice'

export function makeStore() {
  return configureStore({
    reducer: { user: userReducer, userApplications: userApplicationsReducer },
  })
}

const store = makeStore()

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch

export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action<string>>

export default store
