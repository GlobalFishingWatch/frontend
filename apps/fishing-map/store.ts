import type { Action, AnyAction, Middleware, ThunkAction, ThunkDispatch } from '@reduxjs/toolkit'
import { configureStore } from '@reduxjs/toolkit'
import { logoutUserMiddleware } from 'middlewares'
import { createWrapper } from 'next-redux-wrapper'
import { queriesApiMiddlewares } from 'queries'

import connectedRoutes from 'routes/routes'
import { routerQueryMiddleware, routerWorkspaceMiddleware } from 'routes/routes.middlewares'

import { rootReducer } from './reducers'

const {
  middleware: routerMiddleware,
  enhancer: routerEnhancer,
  // initialDispatch,
} = connectedRoutes

// Can't type because GetDefaultMiddlewareOptions type is not exposed by RTK
const defaultMiddlewareOptions: any = {
  // Fix issue with Redux-first-router and RTK (https://stackoverflow.com/questions/59773345/react-toolkit-and-redux-first-router)
  serializableCheck: false,
  immutableCheck: {
    ignoredPaths: [
      // Too big to check for immutability:
      'resources',
      'datasets',
      'dataviews',
    ],
  },
}

const makeStore = () => {
  return configureStore({
    devTools: {
      stateSanitizer: (state: any) => {
        if (!state.resources) return state
        const serializedResources = Object.entries(state.resources).map(([key, value]: any) => [
          key,
          { ...value, data: 'NOT_SERIALIZED' },
        ])

        return {
          ...state,
          resources: Object.fromEntries(serializedResources),
        }
      },
    },
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware(defaultMiddlewareOptions).concat(
        ...queriesApiMiddlewares,
        routerQueryMiddleware,
        routerWorkspaceMiddleware,
        routerMiddleware as Middleware,
        logoutUserMiddleware
      ),
    enhancers: (getDefaultEnhancers) => [routerEnhancer, ...getDefaultEnhancers()] as any,
    // preloadedState,
  })
}

export type AppStore = ReturnType<typeof makeStore>
type TypedDispatch<T> = ThunkDispatch<T, any, AnyAction>

export type AppDispatch = TypedDispatch<RootState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>

// export function useStore(initialState) {
//   const store = useMemo(() => initializeStore(initialState), [initialState])
//   return store
// }
export type RootState = ReturnType<typeof rootReducer>

export const wrapper = createWrapper<AppStore>(makeStore, {
  debug: false,
  serializeState: (state) => {
    return {
      vessel: state.vessel,
      location: { payload: state.location.payload },
    }
  },
  // deserializeState: (state) => JSON.parse(state),
})
