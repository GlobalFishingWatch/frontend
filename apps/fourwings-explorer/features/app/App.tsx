import { useState, useCallback, Fragment } from 'react'
import dynamic from 'next/dynamic'
import { SplitView } from '@globalfishingwatch/ui-components'
import Sidebar from 'features/sidebar/Sidebar'
import useMapInstance from 'features/map/map-context.hooks'
import Modals from 'features/modals/Modals'
import styles from './App.module.css'

const Map = dynamic(() => import(/* webpackChunkName: "Timebar" */ 'features/map/Map'))
const Timebar = dynamic(() => import(/* webpackChunkName: "Timebar" */ 'features/timebar/Timebar'))

const Main = () => {
  return (
    <div className={styles.main}>
      <div className={styles.mapContainer}>
        <Map />
      </div>
      <Timebar />
    </div>
  )
}

function App(): React.ReactElement {
  const [sidebarOpen, setSidebarOpen] = useState(true)
  const map = useMapInstance()

  const onToggle = useCallback(() => {
    setSidebarOpen(!sidebarOpen)
    if (map?.resize) {
      setTimeout(() => {
        map.resize()
      }, 1)
    }
  }, [sidebarOpen, map])

  const asideWidth = '50%'

  return (
    <Fragment>
      <SplitView
        showToggle
        isOpen={sidebarOpen}
        onToggle={onToggle}
        aside={<Sidebar />}
        main={<Main />}
        asideWidth={asideWidth}
        showAsideLabel={'TODO'}
        showMainLabel="Map"
        className="split-container"
      />
      <Modals />
    </Fragment>
  )
}

export default App
