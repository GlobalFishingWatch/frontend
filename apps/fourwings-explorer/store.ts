import { configureStore, ThunkAction, Action, combineReducers } from '@reduxjs/toolkit'
import connectedRoutes from './routes/routes'
import { routerQueryMiddleware } from './routes/routes.middlewares'
import titleReducer from './routes/title.reducer'

const {
  reducer: location,
  middleware: routerMiddleware,
  enhancer: routerEnhancer,
  // initialDispatch,
} = connectedRoutes

const rootReducer = combineReducers({
  location: location,
  title: titleReducer,
})

// Can't type because GetDefaultMiddlewareOptions type is not exposed by RTK
const defaultMiddlewareOptions: any = {
  // Fix issue with Redux-first-router and RTK (https://stackoverflow.com/questions/59773345/react-toolkit-and-redux-first-router)
  serializableCheck: false,
}

const store = configureStore({
  devTools: {
    stateSanitizer: (state: any) => {
      if (!state.resources) return state
      const serializedResources = Object.entries(state.resources).map(([key, value]: any) => [
        key,
        { ...value, data: 'NOT_SERIALIZED' },
      ])

      return {
        ...state,
        resources: Object.fromEntries(serializedResources),
      }
    },
  },
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware(defaultMiddlewareOptions)
      .concat(routerQueryMiddleware)
      .concat(routerMiddleware),
  enhancers: (defaultEnhancers) => [routerEnhancer, ...defaultEnhancers],
})

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = typeof store.dispatch
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>

export default store
