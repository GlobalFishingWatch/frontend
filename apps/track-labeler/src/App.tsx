import React, { Fragment, lazy, Suspense,useState } from 'react'
import { useSelector } from 'react-redux'

import { SplitView } from '@globalfishingwatch/ui-components/split-view'

import Loader from '././features/loader/loader'
import Login from '././features/login/Login'
import { MapboxRefProvider } from '././features/map/map.context'
import { LOGIN } from './routes/routes'
import { getLocationType } from './routes/routes.selectors'

import '@globalfishingwatch/ui-components/base.css'
import '../../../libs/timebar/src/timebar-settings.css'

const Main = lazy(() => import(`././features/main/main.container`))
const Sidebar = lazy(() => import(`././features/sidebar/sidebar.container`))

function App(): React.ReactElement<any> {
  const [sidebarOpen, setSidebarOpen] = useState(true)
  const [animation, setAnimation] = useState(true)
  //const logged = true
  const onToggle = () => {
    setSidebarOpen(!sidebarOpen)
  }

  const locationType = useSelector(getLocationType)

  if (locationType === LOGIN) {
    return <Login />
  }

  const mainContent = (
    <Suspense fallback={<Fragment />}>
      <Main />
    </Suspense>
  )
  const sidebarContent = (
    <Suspense fallback={<Fragment />}>
      <Sidebar />
    </Suspense>
  )

  // Just an effect to init the labeler
  setTimeout(() => setAnimation(false), 2000)
  if (animation) {
    return (
      <Fragment>
        <Loader />
      </Fragment>
    )
  }
  return (
    <MapboxRefProvider>
      <SplitView
        isOpen={sidebarOpen}
        onToggle={onToggle}
        showToggle={false}
        aside={sidebarContent}
        main={mainContent}
        asideWidth="44rem"
        className="split-container"
      />
    </MapboxRefProvider>
  )
}

export default App
