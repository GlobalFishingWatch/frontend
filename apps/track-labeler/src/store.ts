import type {
  Action,
  AnyAction,  Middleware,
  ThunkAction,
  ThunkDispatch} from '@reduxjs/toolkit';
import {
  combineReducers,
  configureStore} from '@reduxjs/toolkit'

import dataviews from '././features/dataviews/dataviews.slice'
import resources from '././features/dataviews/resources.slice'
import mapReducer from '././features/map/map.slice'
import project from '././features/projects/projects.slice'
import rulers from '././features/rulers/rulers.slice'
import timebarReducer from '././features/timebar/timebar.slice'
import userReducer from '././features/user/user.slice'
import selectedtracks from '././features/vessels/selectedTracks.slice'
import vessels from '././features/vessels/vessels.slice'
import connectedRoutes from './routes/routes'
import { routerQueryMiddleware, routerRefreshTokenMiddleware } from './routes/routes.middlewares'

const {
  reducer: location,
  middleware: routerMiddleware,
  enhancer: routerEnhancer,
  // initialDispatch,
} = connectedRoutes

const rootReducer = combineReducers({
  dataviews,
  selectedtracks,
  resources,
  rulers,
  user: userReducer,
  map: mapReducer,
  vessels,
  project,
  timebar: timebarReducer,
  location: location,
})

// Can't type because GetDefaultMiddlewareOptions type is not exposed by RTK
const defaultMiddlewareOptions: any = {
  // Fix issue with Redux-first-router and RTK (https://stackoverflow.com/questions/59773345/react-toolkit-and-redux-first-router)
  serializableCheck: false,
  immutableCheck: {
    ignoredPaths: [
      // Too big to check for immutability:
      'resources',
      'vessels',
    ],
  },
}

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware(defaultMiddlewareOptions).concat([
      routerQueryMiddleware,
      routerRefreshTokenMiddleware,
      routerMiddleware as Middleware,
    ]),
  enhancers: (getDefaultEnhancers) => [routerEnhancer, ...getDefaultEnhancers()] as any,
})

export type RootState = ReturnType<typeof rootReducer>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
type TypedDispatch<T> = ThunkDispatch<T, any, AnyAction>
export type AppDispatch = TypedDispatch<RootState>

export default store
