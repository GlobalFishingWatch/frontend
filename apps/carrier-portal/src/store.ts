import { applyMiddleware, combineReducers, createStore, Middleware } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly'
import { connectRoutes, Options } from 'redux-first-router'
import { stringify, parse } from 'qs'
import { AppActions } from 'types/redux.types'
import {
  REPLACE_URL_PARAMS,
  BASE_URL,
  EVENT_TYPES,
  CONTEXT_LAYERS_IDS,
  DEFAULT_FILTERS,
} from 'data/constants'
import { GraphType, QueryParams, QueryParam } from 'types/app.types'
import { trunc } from 'utils'
import reducers from './redux-modules/reducers'
import routesMap from './routes'

export interface QueryTransformation {
  key: QueryParam
  transformation: (value: any) => any
}

const objectToUrlTransformations: QueryTransformation[] = [
  { key: 'start', transformation: (value: string) => value.split('T')[0] },
  { key: 'end', transformation: (value: string) => value.split('T')[0] },
  { key: 'latitude', transformation: (value: number) => trunc(value) },
  { key: 'longitude', transformation: (value: number) => trunc(value) },
  { key: 'zoom', transformation: (value: number) => trunc(value) },
]

const routesOptions: Options = {
  basename: BASE_URL,
  querySerializer: {
    stringify: (object: QueryParams) => {
      objectToUrlTransformations.forEach(({ key, transformation }) => {
        const value = object[key]
        if (value) {
          object[key] = transformation(value)
        }
      })
      return stringify({ ...object }, { encode: false })
    },
    parse: (string: string) => {
      // This only runs once at the beginning, so running transformations on queryParams selectors
      const { after, before, lat, lng, ...rest } = parse(string, { arrayLimit: 300 })
      // convert legacy params into new ones
      const params = {
        ...rest,
        ...(after && { start: after }),
        ...(before && { end: before }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
      }
      return params
    },
  },
}

// This middleware allows rewriting selectively URL params
// without having to know all of them or redirect by some params
const customRouterQueryMiddleware: Middleware = ({ getState }) => (next) => (
  action: AppActions
) => {
  const routesActions = Object.keys(routesMap)
  // check if action type matches a route type
  const isRouterAction = routesActions.includes(action.type)

  // replaceQuery allows full rewriting of URL params
  if (isRouterAction) {
    const newAction: any = { ...action }

    if (newAction.replaceQuery !== true) {
      const prevQuery = getState().location.query || {}
      newAction.query = {
        ...prevQuery,
        ...newAction.query,
      }
    }
    const { query = {} } = action as any
    const redirect = Object.keys(query)
      .filter((k) => query[k])
      .some((key) => REPLACE_URL_PARAMS.includes(key))
    if (redirect === true) {
      newAction.meta = {
        location: {
          kind: 'redirect',
        },
      }
    }

    // Auto select the graph option depending on the current selection
    let graphType: GraphType = newAction.query.graph
    const isGraphRelatedChange = query.rfmo !== undefined || query.flag !== undefined
    const notHasGraphParams =
      query.graph === undefined && query.start === undefined && query.end === undefined
    // Active the rfmo and eez layers if we any of the areas are selected
    const hasRfmoSelected = newAction.query.rfmo && newAction.query.rfmo.length > 0
    const layers = newAction.query.layer ? [...newAction.query.layer] : DEFAULT_FILTERS.layer
    if (hasRfmoSelected) {
      if (query.rfmo && !layers.includes(CONTEXT_LAYERS_IDS.rfmo)) {
        layers.push(CONTEXT_LAYERS_IDS.rfmo)
      }
    }
    const hasEezSelected = newAction.query.eez && newAction.query.eez.length > 0
    if (hasEezSelected) {
      if (query.eez && !layers.includes(CONTEXT_LAYERS_IDS.eez)) {
        layers.push(CONTEXT_LAYERS_IDS.eez)
      }
    }
    if (layers.length) {
      newAction.query.layer = layers
    }
    if (isGraphRelatedChange && notHasGraphParams && hasRfmoSelected) {
      if (newAction.query.flag && newAction.query.flag.length === 1) {
        graphType = 'time'
      } else {
        graphType =
          newAction.query.eventType === EVENT_TYPES.loitering ? 'loitering-flag' : 'flag-carrier'
      }
    }
    if (graphType) {
      newAction.query.graph = graphType
    }
    next(newAction)
  } else {
    next(action)
  }
}

const { reducer, middleware, enhancer } = connectRoutes(routesMap, routesOptions)
export const rootReducer = combineReducers({ ...reducers, location: reducer })

const middlewares = applyMiddleware(customRouterQueryMiddleware, middleware)
const enhancers = composeWithDevTools(enhancer, middlewares)
const store = createStore(rootReducer, {}, enhancers)

export default store
