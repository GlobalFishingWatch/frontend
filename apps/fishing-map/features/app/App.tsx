import { Fragment, useCallback, useEffect, useLayoutEffect, useState } from 'react'
import { FpsView } from 'react-fps'
import { useTranslation } from 'react-i18next'
import { useSelector } from 'react-redux'
import { ToastContainer } from 'react-toastify'
import MemoryStatsComponent from 'next-react-memory-stats'

import type { Workspace } from '@globalfishingwatch/api-types'
import { Logo, Menu, SplitView } from '@globalfishingwatch/ui-components'

import menuBgImage from 'assets/images/menubg.jpg'
import { FIT_BOUNDS_REPORT_PADDING, ROOT_DOM_ELEMENT } from 'data/config'
import { DEFAULT_WORKSPACE_ID } from 'data/workspaces'
import { useDatasetDrag } from 'features/app/drag-dataset.hooks'
import ErrorBoundary from 'features/app/ErrorBoundary'
import { selectDebugOptions } from 'features/debug/debug.slice'
import { t } from 'features/i18n/i18n'
import { useMapFitBounds } from 'features/map/map-bounds.hooks'
import { useSetMapCoordinates } from 'features/map/map-viewport.hooks'
import AppModals from 'features/modals/Modals'
import { selectReportAreaBounds } from 'features/reports/reports.config.selectors'
import Sidebar from 'features/sidebar/Sidebar'
import { selectIsUserLogged } from 'features/user/selectors/user.selectors'
import { fetchUserThunk } from 'features/user/user.slice'
import { useFitWorkspaceBounds } from 'features/workspace/workspace.hook'
import {
  isWorkspacePasswordProtected,
  selectCurrentWorkspaceId,
  selectWorkspaceCustomStatus,
} from 'features/workspace/workspace.selectors'
import { fetchWorkspaceThunk } from 'features/workspace/workspace.slice'
import { fetchHighlightWorkspacesThunk } from 'features/workspaces-list/workspaces-list.slice'
import {
  HOME,
  PORT_REPORT,
  REPORT,
  SEARCH,
  USER,
  VESSEL,
  VESSEL_GROUP_REPORT,
  WORKSPACE,
  WORKSPACE_REPORT,
  WORKSPACE_SEARCH,
  WORKSPACE_VESSEL,
  WORKSPACES_LIST,
} from 'routes/routes'
import { useBeforeUnload, useLocationConnect, useReplaceLoginUrl } from 'routes/routes.hook'
import {
  selectIsAnyAreaReportLocation,
  selectIsAnySearchLocation,
  selectIsAnyVesselLocation,
  selectIsMapDrawing,
  selectIsVesselGroupReportLocation,
  selectIsVesselLocation,
  selectIsWorkspaceLocation,
  selectLocationType,
  selectWorkspaceId,
} from 'routes/routes.selectors'
import { AsyncReducerStatus } from 'utils/async-slice'

import { selectReadOnly, selectSidebarOpen } from './selectors/app.selectors'
import { useAnalytics } from './analytics.hooks'
import { useAppDispatch } from './app.hooks'
import Main from './Main'

import styles from './App.module.css'

declare global {
  interface Window {
    gtag: any
  }
}

const setMobileSafeVH = () => {
  const vh = window.innerHeight * 0.01
  document.documentElement.style.setProperty('--vh', `${vh}px`)
}

function App() {
  useAnalytics()
  useDatasetDrag()
  useReplaceLoginUrl()
  useBeforeUnload()
  const dispatch = useAppDispatch()
  const sidebarOpen = useSelector(selectSidebarOpen)
  const isMapDrawing = useSelector(selectIsMapDrawing)
  const readOnly = useSelector(selectReadOnly)
  const i18n = useTranslation()
  const { dispatchQueryParams } = useLocationConnect()
  const [menuOpen, setMenuOpen] = useState(false)
  const isWorkspaceLocation = useSelector(selectIsWorkspaceLocation)
  const vesselLocation = useSelector(selectIsVesselLocation)
  const isAreaReportLocation = useSelector(selectIsAnyAreaReportLocation)
  const reportAreaBounds = useSelector(selectReportAreaBounds)
  const isAnySearchLocation = useSelector(selectIsAnySearchLocation)
  const isAnyVesselLocation = useSelector(selectIsAnyVesselLocation)
  const isVesselGroupReportLocation = useSelector(selectIsVesselGroupReportLocation)

  const onMenuClick = useCallback(() => {
    setMenuOpen(true)
  }, [])

  useEffect(() => {
    setMobileSafeVH()
    window.addEventListener('resize', setMobileSafeVH, false)
    return () => window.removeEventListener('resize', setMobileSafeVH)
  }, [])

  const fitMapBounds = useMapFitBounds()
  const setMapCoordinates = useSetMapCoordinates()

  const locationType = useSelector(selectLocationType)
  const currentWorkspaceId = useSelector(selectCurrentWorkspaceId)
  const workspaceCustomStatus = useSelector(selectWorkspaceCustomStatus)
  const userLogged = useSelector(selectIsUserLogged)
  const urlWorkspaceId = useSelector(selectWorkspaceId)
  const fitWorkspaceBounds = useFitWorkspaceBounds()

  // TODO review this as is needed in analysis and workspace but adds a lot of extra logic here
  // probably better to fetch in both components just checking if the workspaceId is already fetched
  const isHomeLocation = locationType === HOME
  const homeNeedsFetch = isHomeLocation && currentWorkspaceId !== DEFAULT_WORKSPACE_ID
  // Checking only when REPORT entrypoint or WORKSPACE_REPORT when workspace is not loaded
  const locationNeedsFetch =
    locationType === REPORT ||
    locationType === VESSEL_GROUP_REPORT ||
    locationType === PORT_REPORT ||
    ((locationType === WORKSPACE_REPORT || isAnyVesselLocation) &&
      currentWorkspaceId !== urlWorkspaceId)
  const hasWorkspaceIdChanged = locationType === WORKSPACE && currentWorkspaceId !== urlWorkspaceId

  useEffect(() => {
    let action: any
    let actionResolved = false
    const fetchWorkspace = async () => {
      action = dispatch(fetchWorkspaceThunk({ workspaceId: urlWorkspaceId as string }))
      const resolvedAction = await action
      if (fetchWorkspaceThunk.fulfilled.match(resolvedAction)) {
        const workspace = resolvedAction.payload as Workspace
        if (!isVesselGroupReportLocation && !isWorkspacePasswordProtected(workspace)) {
          fitWorkspaceBounds(workspace)
        }
      }
      actionResolved = true
    }
    if (
      userLogged &&
      workspaceCustomStatus !== AsyncReducerStatus.Loading &&
      (homeNeedsFetch || locationNeedsFetch || hasWorkspaceIdChanged)
    ) {
      // TODO Can we arrive in a situation where no workspace is ever loaded?
      // In that case static timerange will need to be set manually
      fetchWorkspace()
    }
    return () => {
      if (action && action.abort !== undefined && !actionResolved) {
        action.abort()
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userLogged, homeNeedsFetch, locationNeedsFetch, hasWorkspaceIdChanged])

  useLayoutEffect(() => {
    if (isAreaReportLocation) {
      if (reportAreaBounds) {
        fitMapBounds(reportAreaBounds, { padding: FIT_BOUNDS_REPORT_PADDING })
      } else {
        setMapCoordinates({ latitude: 0, longitude: 0, zoom: 0 })
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  useEffect(() => {
    dispatch(fetchUserThunk({ guest: false }))
  }, [dispatch])

  useEffect(() => {
    dispatch(fetchHighlightWorkspacesThunk())
  }, [dispatch])

  const onToggle = useCallback(() => {
    dispatchQueryParams({ sidebarOpen: !sidebarOpen })
  }, [dispatchQueryParams, sidebarOpen])

  const debugOptions = useSelector(selectDebugOptions)
  const [isReady, setReady] = useState(false)
  useEffect(() => setReady(true), [])
  const showStats = isReady && debugOptions.mapStats === true

  const getSidebarName = useCallback(() => {
    if (locationType === USER) return t('user.title', 'User')
    if (locationType === WORKSPACES_LIST) return t('workspace.title_other', 'Workspaces')
    if (locationType === SEARCH || locationType === WORKSPACE_SEARCH)
      return t('search.title', 'Search')
    if (locationType === VESSEL || locationType === WORKSPACE_VESSEL)
      return t('vessel.title', 'Vessel profile')
    if (isAreaReportLocation) return t('analysis.title', 'Analysis')
    return t('common.layerList', 'Layer list')
  }, [locationType, isAreaReportLocation])

  let asideWidth = '50%'
  if (readOnly) {
    asideWidth = isAreaReportLocation ? '45%' : '34rem'
  } else if (isAnySearchLocation) {
    asideWidth = '100%'
  } else if (isWorkspaceLocation) {
    asideWidth = '39rem'
  }

  if (!i18n.ready) {
    return null
  }

  return (
    <Fragment>
      <a href="https://globalfishingwatch.org" className="print-only">
        <Logo className={styles.logo} />
      </a>
      <div style={{ position: 'fixed', zIndex: 1 }}>
        {showStats && <FpsView top="0" right="8rem" bottom="auto" left="auto" />}
        {showStats && <MemoryStatsComponent corner="topRight" />}
      </div>
      <ErrorBoundary>
        <SplitView
          isOpen={sidebarOpen && !isMapDrawing}
          showToggle={isWorkspaceLocation || vesselLocation}
          onToggle={onToggle}
          aside={<Sidebar onMenuClick={onMenuClick} />}
          main={<Main />}
          asideWidth={asideWidth}
          showAsideLabel={getSidebarName()}
          showMainLabel={t('common.map', 'Map')}
          className={styles.splitContainer}
          asideClassName={styles.aside}
          mainClassName={styles.main}
        />
        {!readOnly && (
          <Menu
            appSelector={ROOT_DOM_ELEMENT}
            bgImage={menuBgImage.src}
            isOpen={menuOpen}
            onClose={() => setMenuOpen(false)}
            activeLinkId="map-data"
          />
        )}
        <AppModals />
        <ToastContainer
          position="top-center"
          className={styles.toastContainer}
          closeButton={false}
        />
      </ErrorBoundary>
    </Fragment>
  )
}

export default App
