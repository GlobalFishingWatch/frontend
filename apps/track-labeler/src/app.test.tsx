import React from 'react'
import { cleanup, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import store from './store'
import App from './App'

jest.mock('@globalfishingwatch/api-client')

afterEach(cleanup)

const _store = store
describe('App', () => {
  test('displays login page', () => {
    const component = render(
      <Provider store={_store}>
        <App />
      </Provider>
    )
    expect(component.container).toMatchSnapshot()
  })
})
