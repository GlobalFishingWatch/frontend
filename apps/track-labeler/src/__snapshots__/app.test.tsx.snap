// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App displays login page 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="textContainer"
    >
      <svg
        class="logoImage"
      >
        gfw-white.svg
      </svg>
      <h1
        class="textTitle"
      >
        Tracks Labeler
      </h1>
      <h2
        class="textSubtitle"
      >
        Only registered users can use this tool. Please log in with your Global Fishing Watch credentials. If you don’t an account, create it here and contact your project lead to include you in this project.
      </h2>
      <a
        class="link"
      >
        LOG IN
      </a>
    </div>
    <p
      class="copyright"
    >
      © Jiri Rezac / Greenpeace
    </p>
  </div>
</div>
`;
